src/plot_api/subroutines.js

```diff
@@ -349,60 +344,52 @@ exports.lsInner = function(gd) {
     return gd._promises.length && Promise.all(gd._promises);
 };
 
-function shouldShowLine(ax, counterAx, side) {
-    return (ax._anchorAxis === counterAx && (ax.mirror || ax.side === side)) ||
-        ax.mirror === 'all' || ax.mirror === 'allticks' ||
-        (ax.mirrors && ax.mirrors[counterAx._id + side]);
+function shouldShowLinesOrTicks(ax, subplot) {
+    return (ax.ticks || ax.showline) &&
+        (subplot === ax._mainSubplot || ax.mirror === 'all' || ax.mirror === 'allticks');
 }
 
-function findCounterAxes(gd, ax, axList) {
-    var counterAxes = [];
-    var anchorAx = ax._anchorAxis;
-    if(anchorAx) {
-        var counterMain = anchorAx._mainAxis;
-        if(counterAxes.indexOf(counterMain) === -1) {
-            counterAxes.push(counterMain);
-            for(var i = 0; i < axList.length; i++) {
-                if(axList[i].overlaying === counterMain._id &&
-                    counterAxes.indexOf(axList[i]) === -1
-                ) {
-                    counterAxes.push(axList[i]);
-                }
-            }
-        }
+/*
+ * should we draw a line on counterAx at this side of ax?
+ * It's assumed that counterAx is known to overlay the subplot we're working on
+ * but it may not be its main axis.
+ */
+function shouldShowLineThisSide(ax, side, counterAx) {
+    // does counterAx get a line at all?
+    if(!counterAx.showline || !counterAx._lw) return false;
+
+    // are we drawing *all* lines for counterAx?
+    if(counterAx.mirror === 'all' || counterAx.mirror === 'allticks') return true;
+
+    var anchorAx = counterAx._anchorAxis;
+
+    // is this a free axis? free axes can only have a subplot side-line with all(ticks)? mirroring
+    if(!anchorAx) return false;
+
+    // in order to handle cases where the user forgot to anchor this axis correctly
+    // (because its default anchor has the same domain on the relevant end)
+    // check whether the relevant position is the same.
+    var sideIndex = alignmentConstants.FROM_BL[side];
+    if(counterAx.side === side) {
+        return anchorAx.domain[sideIndex] === ax.domain[sideIndex];
```

Comment:
This happens fairly often in our mocks, and is super easy for users to do: for example make stacked subplots `xy` and `x2y2`, then leave `y.anchor` out because everything looks right even though it defaults to `x` and not `x2`. Perhaps we should make better default logic (pick the first non-overlaying axis that will actually make a subplot with this axis, or perhaps the first axis that specifies this axis as its own anchor?) but that logic is kinda tricky and nearly circular. Anyway what this does is as long as the subplot you're drawing *looks* like the two axes are anchored together, it will join the lines neatly at the corner.

This wasn't needed before because we generated so many extra overlaying subplots, *one* of them was sure to make a crisp corner...